src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,421 @@
+                'Failed to put off the expiration of the "%s" lock within the specified time.',
+                $key
+            ));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(Key $key)
+    {
+        $filter = array(
+            '_id' => (string) $key,
+            'token' => $this->getUniqueToken($key),
+        );
+
+        $options = array();
+
+        $this->getCollection()->deleteOne($filter, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function exists(Key $key): bool
+    {
+        $filter = array(
+            '_id' => (string) $key,
+            'token' => $this->getUniqueToken($key),
+            'expires_at' => array(
+                '$gt' => $this->createDateTime(),
+            ),
+        );
+
+        $doc = $this->getCollection()->findOne($filter);
+
+        return null !== $doc;
+    }
+
+    /**
+     * Update or Insert a Key.
+     *
+     * @param Key   $key
+     * @param float $ttl Expiry in seconds from now
+     */
+    private function upsert(Key $key, float $ttl)
+    {
+        $now = microtime(true);
+        $token = $this->getUniqueToken($key);
+
+        $filter = array(
+            '_id' => (string) $key,
+            '$or' => array(
+                array(
+                    'token' => $token,
+                ),
+                array(
+                    'expires_at' => array(
+                        '$lte' => $this->createDateTime($now),
+                    ),
+                ),
+            ),
+        );
+
+        $update = array(
+            '$set' => array(
+                '_id' => (string) $key,
+                'token' => $token,
+                'expires_at' => $this->createDateTime($now + $ttl),
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $this->getCollection()->updateOne($filter, $update, $options);
+    }
+
+    /**
+     * Remove expired locks from the collection. This is a fallback for MongoDB
+     * servers prior to version 2.2. For versions 2.2+ self::createTtlIndex is
+     * called instead based on the gcProbablity options.
+     */
+    private function removeExpiredLocks()
+    {
+        $now = microtime(true);
+
+        $filter = array(
+            'expires_at' => array(
+                '$lte' => $this->createDateTime($now),
+            ),
+        );
+
+        $this->getCollection()->deleteMany($filter);
+    }
+
+    /**
+     * @return bool
+     */
+    private function isDuplicateKeyException(WriteException $e): bool
+    {
+        $writeErrors = $e->getWriteResult()->getWriteErrors();
+        if (1 === \count($writeErrors)) {
+            $code = $writeErrors[0]->getCode();
+        } else {
+            $code = $e->getCode();
+        }
+        // Mongo error E11000 - DuplicateKey
+        return 11000 === $code;
+    }
+
+    /**
+     * @return Manager
+     */
+    private function getManager(): Manager
```

Comment:
Can be removed. This method is called once in the class: inside the `getDatabaseVersion` inside a `if cached missed`

Reply:
removed